sound/soc/sof/ipc4-topology.c

```diff
@@ -1341,16 +1360,24 @@ sof_ipc4_prepare_copier_module(struct snd_sof_widget *swidget,
 		copier_data = &ipc4_copier->data;
 		available_fmt = &ipc4_copier->available_fmt;
 		if (dir == SNDRV_PCM_STREAM_CAPTURE) {
+			unsigned int fmt_cfg;
+
 			available_fmt->ref_audio_fmt = available_fmt->out_audio_fmt;
 			ref_audio_fmt_size = sizeof(struct sof_ipc4_audio_format);
 
 			/*
-			 * modify the input params for the dai copier as it only supports
-			 * 32-bit always
+			 * As path walk starts from the capture dai, and copiers can
```

Comment:
@kv2019i this is also very hard to follow. Basically, the reason for limiting the input format to 32 bits in the case of the DAI copier was that we can match and retreive the format based on the output format along from the list of available formats. Now, if you want to change the logic to not limit to 32 bits, then what you need is to change the matching function (which is sof_ipc4_init_audio_fmt) to include 2 references not one. The current function will match only the output format based on pipeline_params. But in your case you need to match the output to match pipeline_params and input to also match the runtime params.

Reply:
Hi @kv2019i I'm a bit confused as to how this works. So you want to allow 16-bit capture in the DAI but here you dont change the ref_audio_fmt or the ref_params? Shouldn't you set the ref_params to fe_params and the ref_audio_fmt to the input_params?

Reply:
@ranj063 More specifically, I'd want to give control of how DAI copier is configured here to topology. The copier can do conversions and in current code, kernel hardcodes this to S32 irrespectively of what is defined in topology for this copier. With this patch, we take the DAI copier output format, and limit the pipeline parameters to its output format.

One confusion is that "fmt" is a mask pointing to pipeline_params and on line 1386 this is copied to ref_params. So we do modify the ref_params which is passed to sof_ipc4_init_audio_fmt().

In sof_ipc4_init_audio_fmt() we have a bit of redundancy as the format selection is already limited, but this is a) the case already as current code overrides fmt to S32, and b) sof_ipc4_init_audio_fmt() is used for all modules so it needs to be more generic.